### PR TITLE
fixing arrays in unions for rust and go

### PR DESCRIPTION
**Pull request by lightyear15**
Linking issue #11

### CODE REVIEW

File: src/generator/rust/mod.rs [hunk 1]
```diff
@@ -97,9 +97,18 @@ static UNION_T: &str = r#"
...
+        {{else}}
+            #[array(var = {{case.ret_type.array_size}})]
+        {{/if}}
+        {{case.value}}(Vec<{{case.ret_type.type_name}}>),
+    {{else}}
+        {{case.value}}({{case.ret_type.type_name}}),
+    {{/if}}
 {{else}}
-  {{case.value}}(()),
+  {{case.value}},
```
Comment (jeff1010322) [inline src/generator/rust/mod.rs:111]: Why did you remove these parenthesis for void types?

This is causing errors when building if a void type is used as the case 0 for the Default implementation:

Example generated and error:
```
#[derive(PartialEq, Clone, Debug, XDROut, XDRIn)]
pub enum ActionCategory {
    NONE,

    CALL(Call),

    UPDATE(Update),
}

impl Default for ActionCategory {
    fn default() -> Self {
        ActionCategory::NONE(())
    }
}
```

```
error[E0618]: expected function, found enum variant `ActionCategory::NONE`
   --> rust/xdr_generated.rs:942:9
    |
933 |     NONE,
    |     ---- `ActionCategory::NONE` defined here
...
942 |         ActionCategory::NONE(())
    |         ^^^^^^^^^^^^^^^^^^^^----
    |         |
    |         call expression requires function
    |
help: `ActionCategory::NONE` is a unit variant, you need to write it without the parenthesis
    |
942 |         ActionCategory::NONE
```

File: src/generator/go/mod.rs [hunk 2]
```diff
@@ -183,7 +183,11 @@ type {{uni.name}} struct{
   {{uni.switch.enum_name}} {{uni.switch.enum_type}}
 {{#each uni.switch.cases as |case|}}
 {{#if (not (isvoid case.ret_type.name))}}
-  {{case.ret_type.name}} *{{case.ret_type.type_name}}
+    {{#if case.ret_type.array_size}}
+        {{case.ret_type.name}} *[]{{case.ret_type.type_name}}
```
Comment (jeff1010322) [inline src/generator/go/mod.rs:187]: I think this needs updated to work with fixed array sizes and for strings.

File: test.x [hunk 3]
```diff
@@ -4,127 +4,146 @@
...
+        Type_Array = 2,
+        Type_Empty = 3
+    };
+
+    union TestUnionSwitchArrayUnion switch (TestUnionSwitchArrayEnum Type)
+    {
+        case Type_Int:
+            hyper anInt;
+        case Type_String:
+            string aString;
```
Comment (jeff1010322) [inline test.x:143]: I tested updating these to fixed lengths and generating and got unexpected results:

```
            hyper anInt<32>;
        case Type_String:
            string aString<50>;
```

Rust:

```
#[derive(PartialEq, Clone, Debug, XDROut, XDRIn)]
pub enum TestUnionSwitchArrayUnion {
    #[array(var = 32)]
    Type_Int(Vec<i64>),

    #[array(var = 50)]
    Type_String(Vec<String>),

    #[array(var = 2147483647)]
    Type_Array(Vec<bool>),

    Type_Empty,
}
```

Go:

```
func NewTestUnionSwitchArrayUnion(aType TestUnionSwitchArrayEnum, value interface{}) (result TestUnionSwitchArrayUnion, err error) {
	result.Type = aType
	switch aType {

	case TestUnionSwitchArrayEnumType_Int:

		tv, ok := value.([]int64)

		if !ok {
			err = fmt.Errorf("invalid value, must be [object]")
			return
		}
		result.AnInt = &tv

	case TestUnionSwitchArrayEnumType_String:

		tv, ok := value.([]string)

		if !ok {
			err = fmt.Errorf("invalid value, must be [object]")
			return
		}
		result.AString = &tv

	case TestUnionSwitchArrayEnumType_Array:

		tv, ok := value.([]bool)

		if !ok {
			err = fmt.Errorf("invalid value, must be [object]")
			return
		}
		result.Bools = &tv

	case TestUnionSwitchArrayEnumType_Empty:

	}
	return
}
```

In both cases it interpreted the String as a an Array of strings instead of a max length string.  And for the Go generated it did not create a fixed length int array.

Comment (lightyear15) [reply test.x:143]: fixed length array has not been implemented for rust either.
Do you want it to be added?

File: src/generator/go/mod.rs [hunk 4]
```diff
@@ -229,7 +241,13 @@ switch {{uni.enum_type}}(aType) {
 {{#if (not (isvoid case.ret_type.name))}}
 // Must{{case.ret_type.name}} retrieves the {{case.ret_type.name}} value from the union,
 // panicing if the value is not set.
-func (u {{uni.name}}) Must{{case.ret_type.name}}() {{case.ret_type.type_name}} {
+    {{#if (eqstr case.ret_type.type_name)}}
+    func (u {{uni.name}}) Must{{case.ret_type.name}}() {{case.ret_type.type_name}} {
```
Comment (jeff1010322) [inline src/generator/go/mod.rs:245]: Adding the conditional here, messes up the comment that is placed.  The function and comment are now separated by a newline that doesn't get fixed with gofmt.  Could this be fixed?

Example:

```
// MustUpdate retrieves the Update value from the union,
// panicing if the value is not set.

func (u ActionCategory) MustUpdate() Update {

	val, ok := u.GetUpdate()

	if !ok {
		panic("arm Update is not set")
	}

	return val
}
```

**Review by jeff1010322 — APPROVED** [review]
